platforms/common/px4_log.cpp

```diff
@@ -124,47 +176,65 @@ __EXPORT void px4_log_modulename(int level, const char *moduleName, const char *
 		};
 		log_message.severity = log_level_table[level];
 
-		unsigned pos = 0;
+		ssize_t pos = snprintf((char *)log_message.text, max_length, __px4__log_modulename_pfmt, module_name);
 
 		va_list argptr;
-
-		pos += snprintf((char *)log_message.text + pos, max_length_pub - pos, __px4__log_modulename_pfmt, moduleName);
 		va_start(argptr, fmt);
-		pos += vsnprintf((char *)log_message.text + pos, max_length_pub - pos, fmt, argptr);
+		pos += vsnprintf((char *)log_message.text + pos, math::max(max_length - pos, (ssize_t)0), fmt, argptr);
 		va_end(argptr);
-		log_message.text[max_length_pub - 1] = 0; //ensure 0-termination
-
+		log_message.text[max_length - 1] = 0; //ensure 0-termination
+		log_message.timestamp = hrt_absolute_time();
 		orb_publish(ORB_ID(log_message), orb_log_message_pub, &log_message);
 	}
-
-#ifdef CONFIG_ARCH_BOARD_PX4_SITL
-	// Without flushing it's tricky to see stdout output when PX4 is started by
-	// a script like for the MAVSDK tests.
-	fflush(out);
-#endif
 }
 
 __EXPORT void px4_log_raw(int level, const char *fmt, ...)
 {
 	FILE *out = stdout;
-	bool use_color = true;
 
 #ifdef __PX4_POSIX
+	bool use_color = true;
 	out = get_stdout(&use_color);
 #endif
 
-#ifndef PX4_LOG_COLORIZED_OUTPUT
-	use_color = false;
-#endif
-
 	if (level >= _PX4_LOG_LEVEL_INFO) {
-		if (use_color) { fputs(__px4_log_level_color[level], out); }
+		static constexpr ssize_t max_length = sizeof(log_message_s::text);
+		char buf[max_length + 1]; // same length as log_message_s::text, but add newline
+		ssize_t pos = 0;
+
+#if defined(PX4_LOG_COLORIZED_OUTPUT)
+
+		if (use_color) {
+			pos += sprintf(buf + pos, "%s", __px4_log_level_color[level]);
+		}
+
+#endif // PX4_LOG_COLORIZED_OUTPUT
 
 		va_list argptr;
 		va_start(argptr, fmt);
-		vfprintf(out, fmt, argptr);
+		pos += vsnprintf(buf + pos, math::max(max_length - pos, (ssize_t)0), fmt, argptr);
 		va_end(argptr);
 
-		if (use_color) { fputs(PX4_ANSI_COLOR_RESET, out); }
+#if defined(PX4_LOG_COLORIZED_OUTPUT)
+
+		if (use_color) {
+			// alway reset color
+			const ssize_t sz = math::min(pos, max_length - (ssize_t)strlen(PX4_ANSI_COLOR_RESET));
+			pos += sprintf(buf + sz, "%s", PX4_ANSI_COLOR_RESET);
+		}
+
+#endif // PX4_LOG_COLORIZED_OUTPUT
+
+		if (pos > max_length) {
```

Comment:
When does this happen? Don't you always check for it, except in the `sprintf`? And why don't you replace the `sprintf` to `snprintf` instead?

Reply:
This is simply preserving the newline from the user input if their string exceeded the max length. If it completely fit then it would be copied forward as is